### PR TITLE
WIP: Rewrite in Typescript

### PR DESCRIPTION
**Pull request by simllll**
Hi all,
this is the initial version of agenda written in typescript. I'm asking for comments, reviews and how we should handle the future of it! 

I tried to refactor everything and add types. This is just an initial version, it should run, but not all tests complete yet.

To clarify what's in here:
- a complete rewrite (I tried keeping git history, but this didn't really work out). Too many files and code logic has changed.
- introduced a DB Adapter Layer, where all the database calls are happening (JobDbRepository) 
- typed everything I could type for now to have type safety in TS and can also export it
- Worked on the test, converted them to ts too and they are using MongoMemoryServer now (this fails on travis right now though, works locally)
- Added eslint&prettier
- Breaking changes (e.g. define(name, processor, options?): options moved from 2nd to 3rd parameter,... )
- touch() can have a progress parameter which is saved to the job

Why?
I think agenda is a great tool, I tried bull and others, but nothing came to the services that agenda offers. But current agenda code base is hard to understand and therefore also hard to maintain.

What now?
Help! Let us fix all tests and improve agenda even more. This is just a inital refactoring, I stumpled on a lot of things that could be improved in the future.

Future?
Exactly, unsure about this.. should we merge it into the agenda repo or should I create a fork (e.g. agenda.ts).. What do you think guys? @niftylettuce @koresar @simison @MichielDeMey @cjolif @rschmukler @OmgImAlexis  & everyone I missed here :)

Todo:
- [x] fix tests
- [ ] describe changes
- [ ] update docs

### CODE REVIEW

**Comment by OmgImAlexis** [comment]
> But agenda as a terrible code base, it's hard to maintain and hard to understand where what happens.

Comments like this are unneeded and don’t add anything to the discussion.

**Comment by simllll** [comment]
> > But agenda as a terrible code base, it's hard to maintain and hard to understand where what happens.
> 
> Comments like this are unneeded and don’t add anything to the discussion.

You are right, sorry I didn't want to confront anyone. I just wanted to add a reasoning why I did this ;) I changed my initial description.

**Comment by OmgImAlexis** [comment]
If you'd like to redo this I'd highly suggest making the changes so it actually tracks correctly in git. With the way you've done this it's seeing all files as deleted and rewritten. This makes tracking back old code a LOT harder.

File: .eslintrc.js [hunk 1]
```diff
@@ -0,0 +1,7 @@
+module.exports = {
+	root: true,
+	extends: ['@hokify'],
+	parserOptions: {
+		project: './tsconfig.eslint.json'
```
Comment (OmgImAlexis) [inline .eslintrc.js:5]: `eslint` can be dropped.

File: .mocharc.jsonc [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+	"diff": true,
+	"require": ["./tshook.js", "source-map-support/register"],
+	"extension": ["js", "ts"],
+	"package": "./package.json",
+	"recursive": true,
+	"reporter": "spec",
+	"slow": 75,
+	"timeout": 5000,
+	"ui": "bdd",
+	"exit": true
```
Comment (OmgImAlexis) [inline .mocharc.jsonc:11]: What's this file added for?

File: .prettierrc [hunk 3]
```diff
@@ -0,0 +1,7 @@
+{
+	"printWidth": 100,
+	"singleQuote": true,
+	"useTabs": true,
```
Comment (OmgImAlexis) [inline .prettierrc:4]: Why was this file added?

**Comment by simllll** [comment]
> If you'd like to redo this I'd highly suggest making the changes so it actually tracks correctly in git. With the way you've done this it's seeing all files as deleted and rewritten. This makes tracking back old code a LOT harder.

I did it as much as it is possible, the thing is this is that most files just really got deleted. I moved some stuff to helpers, but most things are in the class files itself.

File: index.js [hunk 4]
```diff
@@ -1,3 +1,3 @@
-const Agenda = require('./lib/agenda');
+const Agenda = require('./dist');
```
Comment (OmgImAlexis) [inline index.js:1]: Why is a js file being used if this is going to be ts?

**Review by OmgImAlexis — CHANGES_REQUESTED** [review]
Personally I think this is a good idea but the way you've gone about it isn't correct.

**Comment by OmgImAlexis** [comment]
> > If you'd like to redo this I'd highly suggest making the changes so it actually tracks correctly in git. With the way you've done this it's seeing all files as deleted and rewritten. This makes tracking back old code a LOT harder.
> 
> I did it as much as it is possible, the thing is this is that most files just really got deleted. I moved some stuff to helpers, but most things are in the class files itself.

Move a file to `.ts` and commit. Then change the code for the file and commit. This preserves git history. You've broken it completely with this.

File: index.js (hunk 4 above)
Comment (simllll) [reply index.js:1]: well this was the original export file.. 
but I have changed it in package.json anyways, therefore this is an old file.. i just removed it!

File: .prettierrc (hunk 3 above)
Comment (simllll) [reply .prettierrc:4]: I added eslint and prettier to have a well known code style

File: .mocharc.jsonc (hunk 2 above)
Comment (simllll) [reply .mocharc.jsonc:11]: mochars file is a global config file for mocha. It sets stuff like timesouts, but alos hooks like "tshook.js"..this is needed to run typescript files during test.

File: .eslintrc.js (hunk 1 above)
Comment (simllll) [reply .eslintrc.js:5]: eslint was intended, I'm not saying it has to stay, but in general my opinion is that eslint and prettier are a very well known standard nowadays :)

**Comment by simllll** [comment]
> Personally I think this is a good idea but the way you've gone about it isn't correct.

What do you mean by that? The git stuff or the approach itself? Git can be solved, it's not even clear if the community wants to have this in the "original" agenda, so therefore I would like to keep this open for now. And as I said already, it's a pretty huge rewrite, therefore the git (rename) history is limited anyways.

File: .prettierrc (hunk 3 above)
Comment (OmgImAlexis) [reply .prettierrc:4]: This isn't a PR to change the code formatting though.

**Comment by OmgImAlexis** [comment]
> What do you mean by that? The git stuff or the approach itself? 
Make this PR only converting it to typescript, remove all the other changes. The first PR should be just convert all files to ts and enable loose checking. Second PR should be adding types. https://github.com/egoist/tsup should be used for bundling so we get the different bundle we need. It's this kind of things that would have been mentioned if an issue had been opened prior to this.

> therefore the git (rename) history is limited anyways.
Wrong. Go back and look at the rewrite I did from agenda to what it is now. This can be done properly.

File: .prettierrc (hunk 3 above)
Comment (simllll) [reply .prettierrc:4]: > > What do you mean by that? The git stuff or the approach itself?
> 
> Make this PR only converting it to typescript, remove all the other changes. The first PR should be just convert all files to ts and enable loose checking. Second PR should be adding types. https://github.com/egoist/tsup should be used for bundling so we get the different bundle we need. It's this kind of things that would have been mentioned if an issue had been opened prior to this.
> 
> > therefore the git (rename) history is limited anyways.
> 
> Wrong. Go back and look at the rewrite I did from agenda to what it is now. This can be done properly.

I didn't expect a feedback like this, sorry if I somehow insulted you or something. 

It's an initial draft, we can go from here anywhere. I just did this because we use agenda, if you do not agree with my way, this is totally fine. I'm not saying we need to get this into this repo or anything like this. It's just an suggestion.
Code style or not, complete rewrite or not, any other approach (using tsup,.. plain tsc,..) or not,... are all things we can probably solve. If we want to and work together, but we do not need to. It's up to you and the community :) I just wanted to contribute my work, and maybe kick some things off.

Anyways, thanks for your feedback so far.

**Comment by simllll** [comment]
I didn't expect a feedback like this, sorry if I somehow insulted you or something.

It's an initial draft, we can go from here anywhere. I just did this because we use agenda, if you do not agree with my way, this is totally fine. I'm not saying we need to get this into this repo or anything like this. It's just an suggestion.
Code style or not, complete rewrite or not, any other approach (using tsup,.. plain tsc,..) or not,... are all things we can probably solve. If we want to and work together, but we do not need to. It's up to you and the community :) I just wanted to contribute my work, and maybe kick some things off.

Anyways, thanks for your feedback so far.

**Comment by OmgImAlexis** [comment]
> I didn't expect a feedback like this, sorry if I somehow insulted you or something.

It's more you opened a non-draft PR with a bunch of changes without much care for git history. This causes a lot of issues when it come to debugging and I often have to close PRs because of this.

If you'd like to open a new PR with the revised changes then I'd be okay reviewing it but I don't think this PR is in any state to even look at right now with all the changes made to it. Also keep in mind someone is going to have to review every single line as we need to ensure you haven't added anything harmful and with this large of a change that's quite possible even if it's not malicious. This code is used in production by a lot of people.

**Comment by niftylettuce** [comment]
@simllll I'd say to just make a new fork of Agenda and make it something different/new with TS or whatever you like, you'd have full control then, and I'm sure people would be apt to have it linked in the Agenda README like I did with [Bree](https://github.com/breejs/bree).

**Comment by simllll** [comment]
I rebased my PR a little bit, in the PR it's still not showing up, but if you check the commits, you can see the files have been "correctly"  renamed..  I googled it a little bit, and if the changes are too big, somehow git does not show it as renaming of the file anymore.. I guess that's the reason why this is kinda not working like I though it should.

**Comment by OmgImAlexis** [comment]
I’m going to suggest making a new branch from master and adding those commits on. It might fix the commit history.

The files aren’t too big. I tried it locally and renaming all the js to ts tracks correctly.

**Comment by simllll** [comment]
> I’m going to suggest making a new branch from master and adding those commits on. It might fix the commit history.
> 
> The files aren’t too big. I tried it locally and renaming all the js to ts tracks correctly.

Tried it now twice, and also checked another PR preview.. it's not reflecting the renamings :(
See yourself: 
- https://github.com/agenda/agenda/compare/master...hokify:typescript-3?expand=1
- https://github.com/agenda/agenda/compare/master...hokify:typescript-2?expand=1

**Comment by OmgImAlexis** [comment]
It's the file names, why are they different? All file names should be lowercase and only use hyphens.

**Comment by OmgImAlexis** [comment]
`+12,342 −4,890` changes...

This needs to be seperate PRs if you're going to keep working on it.
The file names need to be kept as is with the correct case, etc.
The git history needs to be kept.

**Comment by Uzlopak** [comment]
I reimplemented agendajs in typescript and used mongoose for the connection management. Would be happy to see what you do with the code

**Comment by simllll** [comment]
> I reimplemented agendajs in typescript and used mongoose for the connection management. Would be happy to see what you do with the code

👍  Do you have it hosted on git somewhere?

**Comment by Uzlopak** [comment]
No, I had to modify it extremely for our purposes as we need it to manage it tasks in our microservice architecture (locking of jobs per redis hashmap, Rpc calls instead of spawning of local childprocesses). It is still very close to the original implementation of agendaJS but too heavily changed in specific places.

Do you need some specific help?

**Comment by simllll** [comment]
> No, I had to modify it extremely for our purposes as we need it to manage it tasks in our microservice architecture (locking of jobs per redis hashmap, Rpc calls instead of spawning of local childprocesses). It is still very close to the original implementation of agendaJS but too heavily changed in specific places.
> 
> Do you need some specific help?

Thanks for offering, so far everything is fine :) I'm putting it to a real test and will deploy it to on of our production background servers today. I have some issues with locking currently with the original agenda, and will see if they have solved already. And if not, this would be my next step to add more tests regarding my issue and try to fix it.

The code base itself should be stable now (all tests are passing) and fully "rewritten".

**Comment by OmgImAlexis** [comment]
> `+12,342 −4,890` changes...
> 
> This needs to be seperate PRs if you're going to keep working on it.
> The file names need to be kept as is with the correct case, etc.
> The git history needs to be kept.

Did you just ignore this comment? I’m not going to merge this. This is getting too big. It was too big before you added the latest commits.

I’m in favour of closing this PR. If the changes do land they need to be in smaller focused PRs.

**Comment by koresar** [comment]
Hello there.

Thanks @simllll for the great effort. The changes are indeed too much out of scope of simply rewriting Agenda with TS.
Please, do not incorporate anything new to the project except the actual JS->TS conversion. Otherwise, this becomes a rather dangerous PR.

I see two options at the moment:
* This PR removes all the changes except the JS->TS related stuff.
* We close this PR in favour of a new JS->TS focused PR.

**Comment by koresar** [comment]
@simllll Alternatively, you can publish a new NPM package. I believe that would be the best outcome.

**Comment by Uzlopak** [comment]
I also encourage you to create a new repo for this. Maybe call the new package e.g. agendats.

I mean this project has alot for open PRs, which are disregarded for a long time. I suppose the Alexis wont approve anyway your PR. 

Also I would encourage you to replace the connection management with mongoose logic (but still have mongodb calls without mongoose overhead by using the collection) to throw out issues with replica sets and maybe potential issues with mongodb atlas.

Also locking tasks is not fixed in our code. So yeah... there is an issue with a fix for that, but I will have to check this out. So I guess your code will also have this defect.

**Comment by simison** [comment]
Thanks for working on this @simllll ! Impressive amount of work done here. :-) I'd of course love to see you join maintain Agenda and contribute to common good! 

As you've seen, a big complete rewrite in one PR is certainly causing some friction in folks. I can totally understand that. :-) I share the sentiment voiced by others to go little more slower and strategically about making change happen.

Practically speaking, I'd prolly start from code style changes, TS changes (I believe the git history issue got solved?), and separately with new features, breaking features, etc etc. That'll ensure those changes can be tested and communicated well. That process sure is more work and slower, but in longer run it'll benefit you since you get more eyes to your work and ensure that others are contributing to your code, and not to Agenda separately.

> I also encourage you to create a new repo for this. Maybe call the new package e.g. agendats.

In case of forking, it's likely better find a new name to avoid any confusion.

**Comment by simison** [comment]
@OmgImAlexis  Let's still keep open for the time being to signal that conversation is still ongoing, if you don't mind?

**Comment by simllll** [comment]
Thanks everyone for taking time to read up and add feedback :+1: I really appreciate it. 

To summarize we are still having two options?
a.) fork it - and create a new package
b.) break it up into several PRs.

a.) Would be, at least for me, the least effort - actually there is already a package out called [@hokify/agenda](https://www.npmjs.com/package/@hokify/agenda?g). Downside of this is the "new" (agenda based) project, which puts out another job scheduler to the wild which has to be maintained and spreads resources to two or more projects instead of one. This sounds especially sad if the goal of the two projects would be pretty much the same...
(+) rewrite is easy (git history will be kept so far), but breaking changes would be not a big deal
(+) I can bring in anything in the beginning, and work with PRs later on
(-) Two projects aiming into the same direction
(-) Not so cool for users / community

b.) Would be a great way to keep things moving forward, but it takes way more effort and without active support on splitting it up, I will be doing hard to do find time for it to get this thing going.
(+) We would keep "one" agenda out there.
(-) I can help, but cannot do this by my own.
(-) A complete rewrite is something that is either a breaking major version with bigger changes in it (e.g. 4), or pretty hard to bring into a long-lifed code base by splitting it up. At least that's my feeling about it. But if someone has a roadmap/plan for this, I'm thrilled to make this happen. E.g. like @simison suggested:
1. start from code style changes
2. introduce TS changes
3. introduce db repository / maybe add internal cleanups for filenames etc..
4. new / breaking features
Where currently I have still some major concerns with step 2 and can't really figure out how we should split this up. The rest, as you can see in this current PR, is just a very little part of the code refactoring.

So to decide how to go on, it would be crucial to see if you are up to help and find time to get this going forward (therefore I would say option B) or if you would like me to do this by my own (that is probably option A then).

Anyway, thanks again for your feedback so far!

**Comment by koresar** [comment]
As for the option b), I'm happy to review changes within 24h, if you ping me in Slack.

TS introduction can be gradual without any major rewrites (as far as I'm aware, I'm not a TS expert though).

**Comment by OmgImAlexis** [comment]
https://github.com/agenda/agenda/pull/1129 supersedes this PR. Any new changes should be made AFTER that PR is merged.

**Comment by simison** [comment]
@simllll thanks for the breakdown! Very helpful.

I agree with your sentiment that help with PR reviews and coding would be essential here. I can't help with those unfortunately but happy to help with thinking about order of operations, and generally just connecting dots.

I also try to enable everyone who wants to help with access; hence I've also added you to Agenda org so that it's easier for you to open PRs directly here and not deal with forks if you'd like. :-)

Hope this helps!

**Comment by simllll** [comment]
@simison There is actually one thing I was thinking about and couldn't find the right answer so far, maybe you can shed some light on this:
While refactoring I tried to clean up/understand the locking and processing job logic, basically what is happening is:
1. in `jobProcessing()` it is checked if there is a concurrency free job:
`returnNextConcurrencyFreeJob()`..
2. if so, `runOrRetry` without any arguments is called.. which then just pops up the jobQueue (no additional selection logic) .. the logic of "returnNextConcurrencyFreeJob" is then afterwards kinda duplicated.. 

so either `returnNextCurrencyFreeJob` could check all the conditions that are also checked in "runOrRetry" and pass the job along,
or returnNextCurrencyFreeJob could be kicked at all, and just checks if there is something in the queue?
or why isn't the job that is returned by nextconcurrency free job not just passed to "runOrRetry"?

it is not a big deal, but I just though this could be somehow reasonable and I just don't understand it yet.

Thanks a lot
